### ISSUE TITLE
Apparent bug when using DefinePlugin with module.hot.accept

### ISSUE THREAD

**Issue by justinhelmer** [webpack/webpack#6919]
**Do you want to request a *feature* or report a *bug*?**

bug

**What is the current behavior?**
`module.hot.accept()` causes an error to be thrown when used in conjunction with `DefinePlugin`:

```
Uncaught SyntaxError: missing ) after argument list
```
> **NOTE:** I have seen the other posts about this error being caused by using `--hot` in conjunction with the `HotModuleReplacementPlugin`; this is something different. I am using `webpack-dev-middleware` and not using `webpack-dev-server` with the `--hot` option.

**If the current behavior is a bug, please provide the steps to reproduce.**

For example, assume the following `DefinePlugin` configuration:
```js
// ...
new webpack.DefinePlugin({
  APP_PATH: JSON.stringify('../../../../app.js'), // note the four `../`
}),
// ...
```
...and the following entry point:
```js
console.log(APP_PATH === '../../../../app.js'); // outputs true

const app = require(APP_PATH); // works
app.start();

if (module.hot) {
  module.hot.accept(APP_PATH, () => { /* noop */ });
}
```

The following error will be thrown:

```
Uncaught SyntaxError: missing ) after argument list
```

You will see the compiled output looks like this:
```js
console.log("../../../../app.js" === "../../../../app.js"); // outputs true

var app = __webpack_require__("../../../app.js");
app.start();

if (module.hot) {
  module.hot.accept("../../../app.js""../../../../app.js", function () { /* noop */ });
}
```

**What is the expected behavior?**
When passing the *value* of `APP_PATH` directly, everything works as expected:
```js
console.log(APP_PATH === '../../../../app.js'); // outputs true

const app = require(APP_PATH); // works
app.start();

if (module.hot) {
  module.hot.accept('../../../../app.js'), () => { /* noop */ });
}
```

You will see the compiled output looks like this:
```js
console.log("../../../../app.js" === "../../../../app.js"); // outputs true

var app = __webpack_require__("../../../app.js");
app.start();

if (module.hot) {
  module.hot.accept("../../../../app.js", function () { /* noop */ });
}
```

...and works perfectly.

----

I also just noticed the `__webpack_require__` is receiving `../../../app.js` even though `APP_PATH` is defined as `../../../../app.js` (and it's working), so now I'm really confused.

**Please mention other relevant information such as the browser version, Node.js version, webpack version, and Operating System.**

Chrome
Node 8.9.x
Webpack 3.x
OSX Sierra

Any help would be greatly appreciated.

**Comment by montogeek** [comment]
Thanks for your report. It would be great if you reduce your issue to a small reproducible example. Best put this example into a github repository together with instructions how to get to the problem.

**Comment by justinhelmer** [comment]
I sure can!

https://github.com/justinhelmer/defineplugin-hot-accept-bug

Thanks a lot for the help. Cheers.

**Comment by sokra** [comment]
Send a PR adding `return true` or `return false` before this line
https://github.com/webpack/webpack/blob/f5bd213f4c697259106da528eb9af1d1bb9c8ba3/lib/HotModuleReplacementPlugin.js#L353

**Comment by justinhelmer** [comment]
@sokra -

I assume you mean based on `expr.arguments.length`?

**Comment by justinhelmer** [comment]
Also, I am going to need some direction if you want me to write a test case for this.

**Comment by justinhelmer** [comment]
Lastly, does the same apply to webpack v3? I'm happy to make the PR to master, but I'm stuck on v3 at the moment.

**Comment by sokra** [comment]
You can always return true. It just skips the parser to dig depther into the AST and stop processing here.

**Comment by justinhelmer** [comment]
PR for [webpack@master](https://github.com/webpack/webpack/tree/master) is ready: https://github.com/webpack/webpack/pull/6962

For [webpack@3.11](https://github.com/webpack/webpack/tree/v3.11.0), I get the following output when trying to execute `npm run test` (before making any changes to the source):

```
> webpack@3.11.0 pretest /Volumes/workspace/webpack
> npm run lint-files


> webpack@3.11.0 lint-files /Volumes/workspace/webpack
> npm run lint && npm run beautify-lint && npm run schema-lint


> webpack@3.11.0 lint /Volumes/workspace/webpack
> eslint lib bin hot buildin "test/**/webpack.config.js" "test/binCases/**/test.js" "examples/**/webpack.config.js"


> webpack@3.11.0 beautify-lint /Volumes/workspace/webpack
> beautify-lint "lib/**/*.js" "hot/**/*.js" "bin/**/*.js" "benchmark/*.js" "test/*.js"

Fine. (338 files checked)

> webpack@3.11.0 schema-lint /Volumes/workspace/webpack
> mocha test/*.lint.js --opts test/lint-mocha.opts

  633 passing (73ms)


> webpack@3.11.0 test /Volumes/workspace/webpack
> mocha test/*.test.js test/*.unittest.js --max-old-space-size=4096 --harmony --trace-deprecation --check-leaks

module.js:341
    throw err;
    ^

Error: Cannot find module 'webpack/lib/Chunk'
    at Function.Module._resolveFilename (module.js:339:15)
    at Function.Module._load (module.js:290:25)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Volumes/workspace/webpack/node_modules/extract-text-webpack-plugin/dist/index.js:17:14)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Volumes/workspace/webpack/node_modules/extract-text-webpack-plugin/dist/cjs.js:3:18)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at Object.<anonymous> (/Volumes/workspace/webpack/test/HotTestCases.test.js:8:27)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Module.require (module.js:367:17)
    at require (internal/module.js:16:19)
    at /Volumes/workspace/webpack/node_modules/mocha/lib/mocha.js:231:27
    at Array.forEach (native)
    at Mocha.loadFiles (/Volumes/workspace/webpack/node_modules/mocha/lib/mocha.js:228:14)
    at Mocha.run (/Volumes/workspace/webpack/node_modules/mocha/lib/mocha.js:514:10)
    at Object.<anonymous> (/Volumes/workspace/webpack/node_modules/mocha/bin/_mocha:480:18)
    at Module._compile (module.js:413:34)
    at Object.Module._extensions..js (module.js:422:10)
    at Module.load (module.js:357:32)
    at Function.Module._load (module.js:314:12)
    at Function.Module.runMain (module.js:447:10)
    at startup (node.js:146:18)
    at node.js:404:3
```

This error makes sense because webpack isn't installed as a child dependency of itself. I tried different versions of node/npm thinking the behavior would change, but i always get the same error.

Without a valid control, I can't make the changes to v3 and ensure I am not introducing any breaking changes.

**Comment by justinhelmer** [comment]
I also failed to mention, lint fails before making any changes:
```
webpack/lib/ModuleParseError.js
  18:7  error  Unexpected control character(s) in regular expression: \x01, \x02, \x03, \x04, \x05, \x06, \x07  no-control-regex

✖ 1 problem (1 error, 0 warnings)
```

The problematic source code is here:
```js
// eslint fails on the below line (no-control-regex)
if(/[\0\u0001\u0002\u0003\u0004\u0005\u0006\u0007]/.test(source)) { // binary file
```

I simply circumvented this issue which brought up the second issue (the above comment).

Please let me know how you want me to proceed with these two initial issues that come up by running `npm run test` against the `v3.11` source locally.

**Comment by ooflorent** [comment]
> `Error: Cannot find module 'webpack/lib/Chunk'`

Run the following commands to setup the project:

```
yarn
yarn link
yarn link webpack
```

They are documented into `CONTRIBUTING.md` if you need further help.

> lint fails before making any changes

Add `"no-control-regex": 0` to `.eslintrc`.

**Comment by justinhelmer** [comment]
@ooflorent - Thanks for the info.

It seems after properly setting it up with `yarn`, both the module load and the linting error go away. Not sure how `yarn` fixed the linting but at this point I'm too afraid to ask. 😜Probably something to do with an `eslint` plugin. Anyways, moving on...

---

I'm assuming the [webpack-3](https://github.com/webpack/webpack/blob/webpack-3) branch is what I should create a PR against. However the tip of that branch failed on the very first run of `yarn test`:
```
  9745 passing (8m)
  207 pending
  2 failing

  1) StatsTestCases should print correct stats for define-plugin:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at Timeout.<anonymous> (/Volumes/workspace/webpack/node_modules/mocha/lib/runnable.js:232:19)
      at tryOnTimeout (timers.js:224:11)
      at Timer.listOnTimeout (timers.js:198:5)


  2) StatsTestCases should print correct stats for filter-warnings:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
      at Timeout.<anonymous> (/Volumes/workspace/webpack/node_modules/mocha/lib/runnable.js:232:19)
      at tryOnTimeout (timers.js:224:11)
      at Timer.listOnTimeout (timers.js:198:5)


error Command failed with exit code 2.
```

I ran a fresh clone just to make sure the issue is repeatable. The number of errors fluctuates depending on the test run. For example, on a subsequent test run against a fresh clone, the following errors were reported (output truncated for the sake of brevity):
```
  9743 passing (7m)
  207 pending
  4 failing

  1) StatsTestCases should print correct stats for filter-warnings:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  2) StatsTestCases should print correct stats for preset-mixed-array:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  3) StatsTestCases should print correct stats for preset-none-array:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  4) StatsTestCases should print correct stats for separate-css-bundle:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

error Command failed with exit code 4.
```

On the **third** test run against the _same_ checkout (no fresh clone), the same two errors were reported as the **first** test run against the previous checkout.

So I tried switching `node` / `npm` versions, grabbing a fresh clone, re-installing `yarn` dependencies and running the test suite again. The **fourth** test run got **three** errors (output truncated for the sake of brevity):
```
  9900 passing (7m)
  142 pending
  3 failing

  1) StatsTestCases should print correct stats for commons-plugin-issue-4980:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  2) StatsTestCases should print correct stats for define-plugin:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  3) StatsTestCases should print correct stats for preset-mixed-array:
     Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

error Command failed with exit code 3.
```

---

So it appears that I can run the test suite N number of times and get different results.

Without a valid control, I can't make the changes to v3 and ensure I am not introducing any breaking changes. Should I just ignore these inconsistencies and “push-and-pray”? Or is there some magic you can tell me that will make this issue go away?

I didn’t notice anything like this when I was fixing the bug & running tests in the master branch, so perhaps it’s a known issue that was solved in webpack v4+. I tried searching through the issue queue in the hopes that a previous contributor ran into this, but didn’t find anything (including closed issues).

---

Tested with `yarn@1.6.0` + both `node@5.10.1` / `npm@3.8.3` and `node@8.9.4` / `npm@5.6.0`.
MacBook Pro - OS X 10.12, 2.5 GHz Intel Core i7 w/ 16GB RAM + SSD

**Comment by justinhelmer** [comment]
I have opened PR!7083, despite the above comment. If things fail, it's not due to the code I added. It's due to the issue outlined above.

**Comment by justinhelmer** [comment]
@sokra / @ooflorent - now that #7083 is merged, is there something I need to do in order to get `webpack@3.12` published with these fixes?

Then, I can close this issue.

Thanks

**Comment by justinhelmer** [comment]
Resolved via #7273
